x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealms.java

```diff
@@ -0,0 +1,120 @@
+     * @return Prioritized list of principal claim names (ex: sub, oid, client_id, azp, appid, client_id, email).
+     */
+    public List<String> getPrincipalClaimNames() {
+        return this.principalClaimNames;
+    }
+
+    /**
+     * Call order is request => JwtRealm.token() => JwtRealms.token()
+     * @param threadContext Request headers and parameters
+     * @return JwtAuthenticationToken containing mandatory JWT header, optional client secret, and a realm order cache key
+     */
+    AuthenticationToken token(final ThreadContext threadContext) {
+        // extract value from Authorization header with Bearer scheme prefix
+        final SecureString authenticationParameterValue = JwtUtil.getHeaderValue(
+            threadContext,
+            JwtRealm.HEADER_END_USER_AUTHENTICATION,
+            JwtRealm.HEADER_END_USER_AUTHENTICATION_SCHEME,
+            false
+        );
+        if (authenticationParameterValue == null) {
+            return null;
+        }
+        // extract value from ES-Client-Authentication header with SharedSecret scheme prefix
+        final SecureString clientAuthenticationSharedSecretValue = JwtUtil.getHeaderValue(
+            threadContext,
+            JwtRealm.HEADER_CLIENT_AUTHENTICATION,
+            JwtRealm.HEADER_SHARED_SECRET_AUTHENTICATION_SCHEME,
+            true
+        );
+        return new JwtAuthenticationToken(this.principalClaimNames, authenticationParameterValue, clientAuthenticationSharedSecretValue);
+    }
+
+    public JwtRealm createJwtRealm(
+        final RealmConfig config,
+        final SSLService sslService,
+        final NativeRoleMappingStore nativeRoleMappingStore
+    ) {
+        return this.addRegisteredJwtRealm(new JwtRealm(config, this, sslService, nativeRoleMappingStore));
+    }
+
+    /**
+     * Register a JWT realm.
+     * @param jwtRealm JWT realm to be registered.
+     * @return The added JwtRealm.
+     */
+    public JwtRealm addRegisteredJwtRealm(final JwtRealm jwtRealm) {
+        this.jwtRealms.add(jwtRealm);
+        return jwtRealm;
+    }
+
+    /**
+     * Unregister a JWT realm.
+     * @param jwtRealm JWT realm to be unregistered.
+     * @return The removed JwtRealm.
+     */
+    public JwtRealm removeRegisteredJwtRealm(final JwtRealm jwtRealm) {
+        this.jwtRealms.remove(jwtRealm);
+        return jwtRealm;
+    }
+
+    /**
+     * Unregister all JWT realms.
+     * @return List of removed JwtRealm.
+     */
+    public List<JwtRealm> clearRegisteredJwtRealms() {
+        final List<JwtRealm> jwtRealmList = this.listRegisteredJwtRealms();
+        this.jwtRealms.clear();
+        return jwtRealmList;
+    }
+
+    /**
+     * List all register JWT realms.
+     * @return List of JwtRealm.
+     */
+    public List<JwtRealm> listRegisteredJwtRealms() {
+        return Collections.unmodifiableList(this.jwtRealms);
+    }
```

Comment:
Why do we have these methods?

Reply:
I was thinking of an alternative design. JwtRealm.token() computes realm cache order key without knowing what each JWT realm uses for principal_claims.name. If that was handled by JwtRealms (or perhaps renamed to JwtRealmsManager), it could iterate through each JWT realm and use actual principal_claims.name values, instead of having the prioritized list. In cases were JwtRealm.token() fails for first JWT realm (ex: all prioritized claims missing from JWT), it is wasteful to execute that multiple times, so perhaps something could be done to avoid that extra processing.

Tracking JWT realms is not needed if we stick to the prioritized list. If there was more time right now, I would have liked to discuss that alternative design, but I can remove that code and we can take our time to talk through that idea.

Reply:
Removed tracking of JWT realm instances in that class.